File: 57-insert-interval/insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        if len(intervals) == 0:
            intervals.append(newInterval)
            return intervals
        
        i = 0
        while(i < len(intervals) and intervals[i][0] <= newInterval[0]):
            i+=1
        intervals.insert(i,newInterval)
        res = [intervals[0]]
        for interval in intervals[1:]:
            first = res[-1]
            second = interval
            if first[1] < second[0]:
                res.append(interval)
            else:
                res.pop(-1)
                res.append([min(first[0],second[0]),max(first[1],second[1])])
        return res

File: 57-insert-interval/insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        if (intervals == null || intervals.length == 0) {
            return new int[][]{ newInterval };
        }

        int curr = 0;
        List<int[]> resultList = new ArrayList<>();

        // Add intervals that end before newInterval starts
        while (curr < intervals.length && newInterval[0] > intervals[curr][1]) {
            resultList.add(intervals[curr]);
            curr++;
        }

        // Merge all overlapping intervals
        while (curr < intervals.length && intervals[curr][0] <= newInterval[1]) {
            newInterval[0] = Math.min(newInterval[0], intervals[curr][0]);
            newInterval[1] = Math.max(newInterval[1], intervals[curr][1]);
            curr++;
        }

        // Add the merged interval
        resultList.add(newInterval);

        // Add the rest of the intervals
        while (curr < intervals.length) {
            resultList.add(intervals[curr]);
            curr++;
        }

        return resultList.toArray(new int[0][]);
    }
}
